src/lib/utitlities/synchronization.ts

```diff
@@ -0,0 +1,50 @@
+import type { Invalidator, Readable, Subscriber, Unsubscriber, Writable } from 'svelte/store'
+import type { SnippetStore } from '$lib/utitlities/persistence';
+import type { RemoteSnippetStore } from '$lib/sqlite/sqliterg';
+import { derived, writable } from 'svelte/store';
+
+/**
+ * Represents data state of an in-memory record:
+ *
+ * - `no-remote`: the remote store is not available.
+ * - `synchronized`: the record exists on both local and remote, and the two
+ *   versions do not differ.
+ * - `conflicted`: local version and remote version differs.
+ * */
+type DataState =
+  | 'no-remote'
+  | 'synchronized'
+  | 'conflicted'
+
+type DataStateMap = {[id: string]: DataState}
+
+type SnippetsDataStateStore =
+  Readable<DataStateMap> &
+  {
+    getState(snippetId: string): Promise<DataState>
+  }
+
+export function createSnippetsDataStateStore(
+  localStore: SnippetStore,
+  remoteStore: RemoteSnippetStore,
+): SnippetsDataStateStore {
+  const map: DataStateMap = {}
+  const mapStore = writable(map)
+  derived(
+    [localStore, remoteStore],
+    ([localSnippets, remoteSnippets]) => {
+
+    }
+  )
```

Comment:
I thought quite hard about this. Using the store values directly is not a good choice: we have to rebuild the map from scratch every time a mutation is notified.

I came up with another idea: let the store comes up with an action/log stream that is capable of recreating the whole store from scratch. Some example actions that the stream emits:

- `{"type": "initialization", "snippets": [...]}`
- `{"type": "upsert", "snippet": {id: 1, ...}}`
- `{"type": "upsert", "snippet": {id: 1, ...}}`

Reply:
Theoretically, there is a big performance boost for this way, but I have a question: is it worth it right now? Each store is just a copy of a folder's snippets, which should contains not more than 30 or 40 records. `O(n)` with `n < 100` is not a great performance hog.

We should tackle this in another PR (#40) and focus on making synchronization works for now.